### ISSUE TITLE
Who is using shared_node/shared_layer/Siamese, and what for?

### ISSUE THREAD

**Issue by fchollet** [fchollet/keras#2100]
Quick survey of who is using layer weight sharing, for the purpose of figuring out how to move on with backwards compatibility with the new Keras version. Are you using:
- `layers.core.add_shared_layer`?
- `graph.add_shared_node`?
- `Siamese` layer in a different context?

Thanks.

**Comment by lukedeo** [comment]
I use `graph.add_shared_node` exclusively...

**Comment by farizrahman4u** [comment]
I wrote all three of it, but never used them.

**Comment by EderSantana** [comment]
I use `graph.add_shared_node`

**Comment by fchollet** [comment]
> I wrote all three of it, but never used them.

Really? Then why did you write them?

**Comment by fchollet** [comment]
Current plans are to support `Graph` (although it is deprecated), including `add_shared_node`. `Siamese` is almost certainly getting removed entirely with no legacy support. Unsure yet what to do with `add_shared_layer`, waiting for more replies.

**Comment by jgc128** [comment]
I use `layers.core.add_shared_layer`

**Comment by EderSantana** [comment]
> Really? Then why did you write them?

@fchollet Several users were asking for that. I answered several tickets about how hack a solution for that. @farizrahman4u saved us all writing those layers.

**Comment by farizrahman4u** [comment]
> Really? Then why did you write them?

There were other people in need. There were serious discussions on weight tying going on at that time. also, it was a mental exercise for me.

**Comment by EderSantana** [comment]
> Current plans are to support Graph (although it is deprecated), including add_shared_node. Siamese is almost certainly getting removed entirely with no legacy support. Unsure yet what to do with add_shared_layer, waiting for more replies.

But what about parallel graphs with shared parameters? If we want to do metric learning we need some sort of Siamese support. In other words, we need to apply the same layer twice with different inputs, possibly just changing batch normalizations. 

My PhD research uses that. Here is another example of that: https://devblogs.nvidia.com/parallelforall/understanding-aesthetics-deep-learning/
Although they use Torch for development and Keras for deploy, I believe they could use Keras all the way if provide an easy API to do that.

**Comment by NasenSpray** [comment]
I use `layers.core.add_shared_layer` & `graph.add_shared_node`, but wouldn't mind if they're gone as long as layers/containers keep being `__call__`'able.

**Comment by parag2489** [comment]
I use `add_shared_layer()` in `Sequential()`

**Comment by dbonadiman** [comment]
I use `add_shared_node()` but i don't mind at all rewriting some of mine modules for a more polished API.

**Comment by around1991** [comment]
I use Graph.add_shared_node exclusively. I had to write additional support for getting masks from shared Embedding layers as well, so if the new functional API handles masking inherently, that would save a lot of effort.
